extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -103,13 +109,35 @@ public class RetryableS3OutputStream extends OutputStream
   private boolean error;
   private boolean closed;
 
+  /**
+   * A map to store number of files pending to be uploaded for a given uploadId.
+   */
+  private static final ConcurrentHashMap<String, AtomicInteger> PENDING_FILES = new ConcurrentHashMap<>();
+
+  /**
+   * A map containing the lock for a given uploadId used for notifying the main thread about the completion of
+   * s3.uploadPart() for all chunks for the uploadId, and hence starting the s3.completeMultipartUpload() for
+   * the uploadId.
+   */
+  private static final ConcurrentHashMap<String, Object> FILE_LOCK_MAP = new ConcurrentHashMap<>();
+
+  /**
+   * Semaphore to restrict the maximum number of simultaneous chunks on disk.
+   */
+  private static final int MAX_CONCURRENT_CHUNKS = 10;
+  private static final Semaphore SEMAPHORE = new Semaphore(MAX_CONCURRENT_CHUNKS);
+
+  /**
+   * Threadpool used for uploading the chunks asynchronously.
+   */
+  private static final ExecutorService UPLOAD_EXECUTOR = Executors.newFixedThreadPool(10);
```

Comment:
@cryptoe Similar question here: How should the pool size be configured?

Reply:
This should be dependent on the number of cores. 
Checkout `WorkerMemoryParameters` to figure out how we calculate number of cores. 
I think `Math.max(4,3*numberOfCores)` maybe but we should benchmark.

Reply:
@cryptoe Have made the change. We can benchmark further on this by varying it a bit if needed, but appreciate if you check if the overall changes look good.